File: posnext/posnext/page/restaurant_dashboard/restaurant_dashboard.py
import frappe
from frappe import _
from frappe.utils import today, getdate, now_datetime, get_datetime
import json

@frappe.whitelist()
def get_table_stats():
	"""Get comprehensive table statistics for the dashboard"""
	try:
		# Get total tables
		total_tables = frappe.db.count('Table')

		# Get table status counts
		status_counts = frappe.db.sql("""
			SELECT status, COUNT(*) as count
			FROM `tabTable`
			GROUP BY status
		""", as_dict=True)

		# Initialize counts
		available = 0
		occupied = 0
		reserved = 0

		for status in status_counts:
			if status.status == 'Available':
				available = status.count
			elif status.status == 'Occupied':
				occupied = status.count
			elif status.status == 'Reserved':
				reserved = status.count

		return {
			'total_tables': total_tables,
			'available': available,
			'occupied': occupied,
			'reserved': reserved
		}

	except Exception as e:
		frappe.log_error(f"Error getting table stats: {str(e)}")
		return {
			'total_tables': 0,
			'available': 0,
			'occupied': 0,
			'reserved': 0
		}

@frappe.whitelist()
def get_reservation_stats():
	"""Get reservation statistics for today"""
	try:
		today_date = today()

		# Today's confirmed reservations
		today_reservations = frappe.db.count('Restaurant Reservation', {
			'reservation_date': today_date,
			'status': ['in', ['Confirmed', 'Arrived', 'Seated']]
		})

		# Upcoming reservations for today
		upcoming_reservations = frappe.db.count('Restaurant Reservation', {
			'reservation_date': today_date,
			'status': 'Confirmed',
			'reservation_time': ['>', now_datetime().strftime('%H:%M:%S')]
		})

		return {
			'today_reservations': today_reservations,
			'upcoming_reservations': upcoming_reservations
		}

	except Exception as e:
		frappe.log_error(f"Error getting reservation stats: {str(e)}")
		return {
			'today_reservations': 0,
			'upcoming_reservations': 0
		}

@frappe.whitelist()
def get_occupancy_stats():
	"""Get current occupancy statistics"""
	try:
		# Get current guests seated (from active reservations)
		current_guests = frappe.db.sql("""
			SELECT SUM(party_size) as total_guests
			FROM `tabRestaurant Reservation`
			WHERE status = 'Seated'
			AND reservation_date = %s
		""", (today(),), as_dict=True)[0].total_guests or 0

		# Get table utilization percentage
		total_capacity = frappe.db.sql("""
			SELECT SUM(seating_capacity) as total_capacity
			FROM `tabTable`
		""", as_dict=True)[0].total_capacity or 1

		occupied_capacity = frappe.db.sql("""
			SELECT SUM(t.seating_capacity) as occupied_capacity
			FROM `tabTable` t
			INNER JOIN `tabRestaurant Reservation` rr ON rr.assigned_table = t.name
			WHERE rr.status = 'Seated'
			AND rr.reservation_date = %s
		""", (today(),), as_dict=True)[0].occupied_capacity or 0

		table_utilization = round((occupied_capacity / total_capacity) * 100, 1)

		return {
			'current_guests': current_guests,
			'table_utilization': table_utilization
		}

	except Exception as e:
		frappe.log_error(f"Error getting occupancy stats: {str(e)}")
		return {
			'current_guests': 0,
			'table_utilization': 0
		}

@frappe.whitelist()
def get_revenue_stats():
	"""Get revenue statistics for today and month"""
	try:
		today_date = today()
		current_month = getdate(today_date).strftime('%Y-%m')

		# Today's revenue from POS invoices
		today_revenue = frappe.db.sql("""
			SELECT COALESCE(SUM(grand_total), 0) as revenue
			FROM `tabPOS Invoice`
			WHERE DATE(posting_date) = %s
			AND docstatus = 1
		""", (today_date,), as_dict=True)[0].revenue

		# Current month revenue
		month_revenue = frappe.db.sql("""
			SELECT COALESCE(SUM(grand_total), 0) as revenue
			FROM `tabPOS Invoice`
			WHERE DATE_FORMAT(posting_date, '%%Y-%%m') = %s
			AND docstatus = 1
		""", (current_month,), as_dict=True)[0].revenue

		return {
			'today_revenue': round(today_revenue, 2),
			'month_revenue': round(month_revenue, 2)
		}

	except Exception as e:
		frappe.log_error(f"Error getting revenue stats: {str(e)}")
		return {
			'today_revenue': 0.0,
			'month_revenue': 0.0
		}

@frappe.whitelist()
def get_table_grid():
	"""Get table grid data for visual display"""
	try:
		tables = frappe.db.sql("""
			SELECT name, table_name, seating_capacity, status
			FROM `tabTable`
			ORDER BY table_name
		""", as_dict=True)

		# Enhance with reservation info for reserved tables
		for table in tables:
			if table.status == 'Reserved':
				reservation = frappe.db.sql("""
					SELECT customer_name, reservation_time
					FROM `tabRestaurant Reservation`
					WHERE assigned_table = %s
					AND reservation_date = %s
					AND status = 'Confirmed'
					ORDER BY reservation_time
					LIMIT 1
				""", (table.name, today()), as_dict=True)

				if reservation:
					table.customer_name = reservation[0].customer_name
					table.reservation_time = reservation[0].reservation_time

		return tables

	except Exception as e:
		frappe.log_error(f"Error getting table grid: {str(e)}")
		return []

@frappe.whitelist()
def get_today_reservations():
	"""Get today's reservations for the dashboard"""
	try:
		today_date = today()

		reservations = frappe.db.sql("""
			SELECT name, customer_name, party_size, reservation_time,
				   assigned_table, status, phone
			FROM `tabRestaurant Reservation`
			WHERE reservation_date = %s
			AND status IN ('Confirmed', 'Arrived', 'Seated')
			ORDER BY reservation_time
			LIMIT 20
		""", (today_date,), as_dict=True)

		# Format reservation time for display
		for res in reservations:
			if res.reservation_time:
				res.reservation_time = res.reservation_time.strftime('%I:%M %p')

		return reservations

	except Exception as e:
		frappe.log_error(f"Error getting today's reservations: {str(e)}")
		return []

@frappe.whitelist()
def get_recent_activity():
	"""Get recent restaurant activity for dashboard"""
	try:
		# Recent reservations
		recent_reservations = frappe.db.sql("""
			SELECT 'Reservation' as type, customer_name as description,
				   creation as timestamp, status
			FROM `tabRestaurant Reservation`
			WHERE DATE(creation) >= DATE_SUB(CURDATE(), INTERVAL 7 DAY)
			ORDER BY creation DESC
			LIMIT 10
		""", as_dict=True)

		# Recent POS invoices
		recent_invoices = frappe.db.sql("""
			SELECT 'Sale' as type, customer as description,
				   posting_date as timestamp, grand_total as amount
			FROM `tabPOS Invoice`
			WHERE DATE(posting_date) >= DATE_SUB(CURDATE(), INTERVAL 7 DAY)
			AND docstatus = 1
			ORDER BY posting_date DESC
			LIMIT 10
		""", as_dict=True)

		# Combine and sort by timestamp
		activities = recent_reservations + recent_invoices
		activities.sort(key=lambda x: x.timestamp, reverse=True)

		# Format for display
		for activity in activities[:15]:  # Limit to 15 most recent
			if activity.type == 'Reservation':
				activity.description = f"Reservation: {activity.description} ({activity.status})"
			else:
				activity.description = f"Sale: {activity.description} - ${activity.amount}"
			activity.timestamp = activity.timestamp.strftime('%m/%d %I:%M %p')

		return activities[:15]

	except Exception as e:
		frappe.log_error(f"Error getting recent activity: {str(e)}")
		return []

@frappe.whitelist()
def get_menu_performance():
	"""Get menu item performance statistics"""
	try:
		# Top selling items today
		top_items = frappe.db.sql("""
			SELECT mi.item_name, COUNT(*) as order_count,
				   SUM(poi.amount) as total_revenue
			FROM `tabPOS Invoice Item` poi
			INNER JOIN `tabMenu Item` mi ON poi.item_code = mi.name
			INNER JOIN `tabPOS Invoice` pi ON poi.parent = pi.name
			WHERE DATE(pi.posting_date) = CURDATE()
			AND pi.docstatus = 1
			GROUP BY mi.item_name
			ORDER BY order_count DESC
			LIMIT 10
		""", as_dict=True)

		return top_items

	except Exception as e:
		frappe.log_error(f"Error getting menu performance: {str(e)}")
		return []

File: posnext/posnext/page/restaurant_dashboard/restaurant_dashboard.js
frappe.pages['restaurant-dashboard'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Restaurant Dashboard'),
		single_column: true
	});

	// Add custom CSS for modern styling
	frappe.dom.add_style(`
		.restaurant-dashboard {
			padding: 20px;
			background: linear-gradient(135deg, #ff9a9e 0%, #fecfef 100%);
			min-height: 100vh;
		}
		.dashboard-cards {
			display: grid;
			grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
			gap: 20px;
			margin-bottom: 30px;
		}
		.dashboard-card {
			background: white;
			border-radius: 15px;
			padding: 25px;
			box-shadow: 0 10px 30px rgba(0,0,0,0.1);
			transition: transform 0.3s ease, box-shadow 0.3s ease;
			border-left: 5px solid #ff9a9e;
		}
		.dashboard-card:hover {
			transform: translateY(-5px);
			box-shadow: 0 15px 40px rgba(0,0,0,0.15);
		}
		.card-title {
			font-size: 18px;
			font-weight: 600;
			color: #333;
			margin-bottom: 15px;
			display: flex;
			align-items: center;
		}
		.card-title i {
			margin-right: 10px;
			color: #ff9a9e;
		}
		.metric-value {
			font-size: 32px;
			font-weight: bold;
			color: #ff9a9e;
			margin-bottom: 5px;
		}
		.metric-label {
			color: #666;
			font-size: 14px;
		}
		.quick-actions {
			display: grid;
			grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
			gap: 15px;
			margin-bottom: 30px;
		}
		.action-btn {
			background: white;
			border: 2px solid #ff9a9e;
			border-radius: 10px;
			padding: 15px;
			text-align: center;
			cursor: pointer;
			transition: all 0.3s ease;
			text-decoration: none;
			display: block;
			color: #ff9a9e;
			font-weight: 600;
		}
		.action-btn:hover {
			background: #ff9a9e;
			color: white;
			transform: translateY(-2px);
			box-shadow: 0 5px 15px rgba(255, 154, 158, 0.3);
		}
		.table-grid {
			display: grid;
			grid-template-columns: repeat(auto-fill, minmax(120px, 1fr));
			gap: 15px;
			margin-bottom: 30px;
		}
		.table-card {
			background: white;
			border-radius: 10px;
			padding: 15px;
			text-align: center;
			box-shadow: 0 5px 15px rgba(0,0,0,0.1);
			transition: all 0.3s ease;
		}
		.table-card.available {
			border-left: 4px solid #28a745;
		}
		.table-card.occupied {
			border-left: 4px solid #dc3545;
		}
		.table-card.reserved {
			border-left: 4px solid #ffc107;
		}
		.table-card:hover {
			transform: translateY(-2px);
			box-shadow: 0 8px 25px rgba(0,0,0,0.15);
		}
		.table-name {
			font-weight: bold;
			margin-bottom: 5px;
		}
		.table-capacity {
			font-size: 12px;
			color: #666;
		}
		.reservations-list {
			background: white;
			border-radius: 15px;
			padding: 25px;
			box-shadow: 0 10px 30px rgba(0,0,0,0.1);
		}
		.reservation-item {
			padding: 10px 0;
			border-bottom: 1px solid #eee;
			display: flex;
			justify-content: space-between;
			align-items: center;
		}
		.reservation-item:last-child {
			border-bottom: none;
		}
		.reservation-time {
			font-weight: bold;
			color: #ff9a9e;
		}
		.reservation-details {
			flex: 1;
			margin-left: 15px;
		}
		.status-badge {
			padding: 3px 8px;
			border-radius: 12px;
			font-size: 12px;
			font-weight: 600;
		}
		.status-confirmed { background: #d4edda; color: #155724; }
		.status-arrived { background: #fff3cd; color: #856404; }
		.status-seated { background: #d1ecf1; color: #0c5460; }
	`);

	wrapper.innerHTML = `
		<div class="restaurant-dashboard">
			<div class="quick-actions">
				<a href="#posnext" class="action-btn">
					<i class="fa fa-cutlery"></i><br>
					New Order
				</a>
				<a href="#List/Restaurant%20Reservation" class="action-btn">
					<i class="fa fa-calendar"></i><br>
					Reservations
				</a>
				<a href="#List/Table" class="action-btn">
					<i class="fa fa-table"></i><br>
					Table Management
				</a>
				<a href="#List/KOT" class="action-btn">
					<i class="fa fa-list-alt"></i><br>
					Kitchen Orders
				</a>
				<a href="#List/Menu%20Item" class="action-btn">
					<i class="fa fa-book"></i><br>
					Menu Management
				</a>
				<a href="#List/POS%20Invoice" class="action-btn">
					<i class="fa fa-file-text-o"></i><br>
					Billing
				</a>
			</div>

			<div class="dashboard-cards">
				<div class="dashboard-card">
					<div class="card-title">
						<i class="fa fa-table"></i>
						Table Status
					</div>
					<div id="table-stats">
						<div class="metric-value">Loading...</div>
						<div class="metric-label">Total Tables</div>
					</div>
				</div>

				<div class="dashboard-card">
					<div class="card-title">
						<i class="fa fa-calendar-check-o"></i>
						Today's Reservations
					</div>
					<div id="reservation-stats">
						<div class="metric-value">Loading...</div>
						<div class="metric-label">Confirmed Reservations</div>
					</div>
				</div>

				<div class="dashboard-card">
					<div class="card-title">
						<i class="fa fa-users"></i>
						Current Occupancy
					</div>
					<div id="occupancy-stats">
						<div class="metric-value">Loading...</div>
						<div class="metric-label">Guests Seated</div>
					</div>
				</div>

				<div class="dashboard-card">
					<div class="card-title">
						<i class="fa fa-dollar"></i>
						Today's Revenue
					</div>
					<div id="revenue-stats">
						<div class="metric-value">Loading...</div>
						<div class="metric-label">Total Revenue</div>
					</div>
				</div>
			</div>

			<div class="table-grid" id="table-grid">
				<!-- Tables will be loaded here -->
			</div>

			<div class="reservations-list">
				<h3 style="margin-bottom: 20px; color: #333;">
					<i class="fa fa-clock-o"></i> Today's Reservations
				</h3>
				<div id="reservations-list">
					<div class="reservation-item">
						<span>Loading reservations...</span>
					</div>
				</div>
			</div>
		</div>
	`;

	// Load dashboard data
	load_restaurant_dashboard_data(wrapper);
};

function load_restaurant_dashboard_data(wrapper) {
	// Load table statistics
	frappe.call({
		method: 'posnext.posnext.page.restaurant_dashboard.restaurant_dashboard.get_table_stats',
		callback: function(r) {
			if (r.message) {
				let stats = r.message;
				$(wrapper).find('#table-stats').html(`
					<div class="metric-value">${stats.total_tables}</div>
					<div class="metric-label">Total Tables</div>
					<div style="margin-top: 10px; font-size: 12px;">
						<span style="color: #28a745;">${stats.available} Available</span> |
						<span style="color: #dc3545;">${stats.occupied} Occupied</span> |
						<span style="color: #ffc107;">${stats.reserved} Reserved</span>
					</div>
				`);
			}
		}
	});

	// Load reservation statistics
	frappe.call({
		method: 'posnext.posnext.page.restaurant_dashboard.restaurant_dashboard.get_reservation_stats',
		callback: function(r) {
			if (r.message) {
				let stats = r.message;
				$(wrapper).find('#reservation-stats').html(`
					<div class="metric-value">${stats.today_reservations}</div>
					<div class="metric-label">Confirmed Reservations</div>
					<div style="margin-top: 10px; font-size: 12px;">
						${stats.upcoming_reservations} upcoming today
					</div>
				`);
			}
		}
	});

	// Load occupancy statistics
	frappe.call({
		method: 'posnext.posnext.page.restaurant_dashboard.restaurant_dashboard.get_occupancy_stats',
		callback: function(r) {
			if (r.message) {
				let stats = r.message;
				$(wrapper).find('#occupancy-stats').html(`
					<div class="metric-value">${stats.current_guests}</div>
					<div class="metric-label">Guests Seated</div>
					<div style="margin-top: 10px; font-size: 12px;">
						${stats.table_utilization}% table utilization
					</div>
				`);
			}
		}
	});

	// Load revenue statistics
	frappe.call({
		method: 'posnext.posnext.page.restaurant_dashboard.restaurant_dashboard.get_revenue_stats',
		callback: function(r) {
			if (r.message) {
				let stats = r.message;
				$(wrapper).find('#revenue-stats').html(`
					<div class="metric-value">$${stats.today_revenue}</div>
					<div class="metric-label">Today's Revenue</div>
					<div style="margin-top: 10px; font-size: 12px;">
						$${stats.month_revenue} this month
					</div>
				`);
			}
		}
	});

	// Load table grid
	frappe.call({
		method: 'posnext.posnext.page.restaurant_dashboard.restaurant_dashboard.get_table_grid',
		callback: function(r) {
			if (r.message) {
				let tables = r.message;
				let table_html = tables.map(table => `
					<div class="table-card ${table.status.toLowerCase()}" onclick="open_table('${table.name}')">
						<div class="table-name">${table.table_name}</div>
						<div class="table-capacity">${table.seating_capacity} seats</div>
						<div style="margin-top: 5px; font-size: 11px; color: #666;">${table.status}</div>
					</div>
				`).join('');

				$(wrapper).find('#table-grid').html(table_html);
			}
		}
	});

	// Load today's reservations
	frappe.call({
		method: 'posnext.posnext.page.restaurant_dashboard.restaurant_dashboard.get_today_reservations',
		callback: function(r) {
			if (r.message) {
				let reservations = r.message;
				let reservation_html = reservations.map(res => `
					<div class="reservation-item">
						<div class="reservation-time">${res.reservation_time}</div>
						<div class="reservation-details">
							<strong>${res.customer_name}</strong> - Party of ${res.party_size}
							${res.assigned_table ? `<br>Table: ${res.assigned_table}` : ''}
						</div>
						<span class="status-badge status-${res.status.toLowerCase().replace(' ', '')}">${res.status}</span>
					</div>
				`).join('');

				$(wrapper).find('#reservations-list').html(reservation_html || '<div class="reservation-item"><span>No reservations for today</span></div>');
			}
		}
	});
}

function open_table(table_name) {
	// Open table details or quick actions
	frappe.set_route('Form', 'Table', table_name);
}